File: Android_Intents_ActivityLifeCycle/app/src/main/java/com/example/imageveiwer3/MainActivity.java
package com.example.imageveiwer3;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.ArrayList;

public class MainActivity<imageList> extends AppCompatActivity {

    static final int REQUESTCODE = 0;
    private Context context = this;
    private LinearLayout linearLayout;
    private Button getPicButton;
    ArrayList<ImageViewerModel> imageList = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        Log.i("ActivityLifecycle",getLocalClassName() +  " - onCreate");
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        linearLayout = findViewById(R.id.inter_LinearLayout);
        getPicButton = findViewById(R.id.button_get_pic);
        getPicButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent pickImage = new Intent(Intent.ACTION_PICK);
                pickImage.setType("image/*");
                pickImage.setAction(Intent.ACTION_GET_CONTENT);
                startActivityForResult(pickImage,REQUESTCODE);
            }
        });


    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if (resultCode == RESULT_OK && REQUESTCODE == requestCode){
            if (data != null){
                Uri imageUri = data.getData();
                ImageViewerModel imageViewerModel = new ImageViewerModel(imageUri.toString(), "imageUri");
                imageList.add(imageViewerModel);
                generateTextView(Integer.toString(imageList.size()), imageList.indexOf(imageViewerModel));
            }

        }
    }

    public TextView generateTextView( String text, final int index){
       final TextView listedPicture = new TextView(context);

        listedPicture.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                listedPicture.setTag(imageList);
                Intent sentImage = new  Intent(context,ImageViewerDetails.class);
                ImageViewerModel imageViewerModel = imageList.get(index);
                sentImage.putExtra("sentImage",imageViewerModel);
                startActivity(sentImage);
            }
        });

        listedPicture.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT,ViewGroup.LayoutParams.WRAP_CONTENT));
        listedPicture.setText(text);
        listedPicture.setTextSize(25);
        listedPicture.setPadding(15,15,15,15);
        linearLayout.addView(listedPicture);

        return listedPicture;

    }


    @Override
    protected void onStart() {
        super.onStart();
        Log.i("ActivityLifecycle",getLocalClassName() +  " - onStart");

    }

    @Override
    protected void onResume() {
        super.onResume();
        Log.i("ActivityLifecycle",getLocalClassName() +  " - onResume");
    }

    @Override
    protected void onPause() {
        super.onPause();
        Log.i("ActivityLifecycle",getLocalClassName() +  " - onPause");
    }

    @Override
    protected void onStop() {
        super.onStop();
        Log.i("ActivityLifecycle",getLocalClassName() +  " - onStop");
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Log.i("ActivityLifecycle",getLocalClassName() +  " - onDestroy");
    }


}

File: app/src/main/java/com/lambdaschool/favoritepicturesgallery/MainActivity.kt
package com.lambdaschool.favoritepicturesgallery

import android.content.Context
import android.content.Intent
import android.graphics.Typeface
import android.net.Uri
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.util.TypedValue
import android.view.View
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import kotlinx.android.synthetic.main.activity_main.*

import java.util.ArrayList

class MainActivity : AppCompatActivity() {

    internal var imageList: ArrayList<ImageData> = ArrayList()
    private var layoutManager: LinearLayoutManager? = null
    private var listAdapter: ImageListAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Toast.makeText(this, "Lifecycle - onCreate", Toast.LENGTH_SHORT).show()
        Log.i(javaClass.simpleName, "Lifecycle-onCreate")
        setContentView(R.layout.activity_main)
        val context = this

        imageList = ArrayList()

        button_add.setOnClickListener {
            val intent = Intent(Intent.ACTION_GET_CONTENT)
            intent.type = "image/*"
            if (intent.resolveActivity(packageManager) != null) {
                startActivityForResult(intent, REQUEST_IMAGE_GET)
            }
        }
        Log.i("Activity_State_Log_Data", imageList.toString())


        // use this setting to improve performance if you know that changes
        // in content do not change the layout size of the RecyclerView
        layout_list.setHasFixedSize(true)

        // use a linear layout manager
        layoutManager = LinearLayoutManager(this)
        layout_list.layoutManager = layoutManager

        // specify an adapter (see also next example)
        listAdapter = ImageListAdapter(imageList, this)
        layout_list.adapter = listAdapter

    }

    private fun refreshListView() {
        listAdapter!!.notifyDataSetChanged()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == REQUEST_IMAGE_GET && resultCode == RESULT_OK) {
            val fullPhotoUri = data!!.data
            if (fullPhotoUri != null) {
                imageList.add(ImageData(fullPhotoUri))
                listAdapter!!.notifyItemInserted(imageList.size - 1)
            }
        } else if (requestCode == EDIT_IMAGE_REQUEST && resultCode == RESULT_OK) {
            // Make sure the request was successful
            val returnedData = data!!.getSerializableExtra("object") as ImageData

            for (i in imageList.indices) {
                if (imageList[i].fileUriString == returnedData.fileUriString) {
                    imageList[i] = returnedData
                }
            }

        }
        refreshListView()

    }

    companion object {

        internal const val REQUEST_IMAGE_GET = 1
        internal const val EDIT_IMAGE_REQUEST = 2
    }
    override fun onRestart() {
        Toast.makeText(this, "Lifecycle - onRestart", Toast.LENGTH_SHORT).show()
        Log.i(javaClass.simpleName, "Lifecycle-onRestart")
        super.onRestart()
    }
    override fun onDestroy() {
        Toast.makeText(this, "Lifecycle - onDestroy", Toast.LENGTH_SHORT).show()
        Log.i(javaClass.simpleName, "Lifecycle-onDestroy")
        super.onDestroy()
    }
    override fun onStop() {
        super.onStop()
        Toast.makeText(this, "Lifecycle - onStop", Toast.LENGTH_SHORT).show()
        Log.i(javaClass.simpleName, "Lifecycle-onStop")

    }
    override fun onPause() {
        super.onPause()
        Toast.makeText(this, "Lifecycle - onPause", Toast.LENGTH_SHORT).show()
        Log.i(javaClass.simpleName, "Lifecycle-onPause")

    }
    override fun onResume() {
        super.onResume()
        Toast.makeText(this, "Lifecycle - onResume", Toast.LENGTH_SHORT).show()
        Log.i(javaClass.simpleName, "Lifecycle-onResume")

    }
    override fun onStart() {
        super.onStart()
        Toast.makeText(this, "Lifecycle - onStart", Toast.LENGTH_SHORT).show()
        Log.i(javaClass.simpleName, "Lifecycle-onStart")
    }


}
